plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py

```diff
@@ -147,38 +179,71 @@ def setup(
         )
         self.sweep_dir = config.hydra.sweep.dir
 
+    def _get_directions(self) -> List[str]:
+        if isinstance(self.direction, MutableSequence):
+            return [d.name if isinstance(d, Direction) else d for d in self.direction]
+        elif isinstance(self.direction, str):
+            return [self.direction]
+        return [self.direction.name]
+
+    def _configure_trials(
+        self,
+        trials: List[Trial],
+        search_space: Dict[str, BaseDistribution],
+        fixed_params: Dict[str, Any],
+    ) -> Sequence[Sequence[str]]:
+        overrides = []
+        for trial in trials:
+            for param_name, distribution in search_space.items():
+                trial._suggest(param_name, distribution)
+            for param_name, value in fixed_params.items():
+                trial.set_user_attr(param_name, value)
+
+            if self.custom_search_space:
+                self.custom_search_space(self.config, trial)
+
+            overlap = trial.params.keys() & trial.user_attrs
+            assert len(overlap) == 0, (
+                "Overlapping fixed parameters and search space parameters found!"
+                f"Overlapping parameters: {list(overlap)}"
+            )
+            params = dict(trial.params)
+            params.update(fixed_params)
+
+            def create_override(name: str, value: Any) -> str:
+                override = f"{name}={value}"
+                if name in fixed_params:
+                    return override
+
+                assert self.config is not None
+                if OmegaConf.select(self.config, name) is not None:
+                    return override
+
+                # If the added a trial parameter using custom search space that is not part of the config yet,
+                # we have to add it as a new config value
+                return f"+{override}"
```

Comment:
I am still not comfortable with this; it breaks consistency with Hydra's command-line API as well as with the precedent from PR #1886.
Hydra should fail fast if someone tries to set a parameter that does not exist in the config ... *unless* if the user explicitly adds a plus `+` to signal that adding the new parameter is intended.

This being said, I agree with you that it should be possible to use `custom_search_space` to add a new parameter to the config. Here is my proposal:
The user's custom search space function should call `trial.suggest_float` using a plus symbol for the name of the suggested parameter. For example:
```python
# custom-search-space-objective.py

...

def configure(cfg: DictConfig, trial: Trial) -> None:
    x_value = trial.params["x"]
    trial.suggest_float(
        "z",
        x_value - cfg.max_z_difference_from_x,
        x_value + cfg.max_z_difference_from_x,
    )
    trial.suggest_float("+w", 0.0, 1.0)  # note +w here, not w

...
```

Reply:
Changed it accordingly